libs/entitlement/asm-provider/src/main/java/org/elasticsearch/entitlement/instrumentation/impl/InstrumentationServiceImpl.java

```diff
@@ -48,22 +50,108 @@ public MethodVisitor visitMethod(
                 String[] exceptions
             ) {
                 var mv = super.visitMethod(access, checkerMethodName, checkerMethodDescriptor, signature, exceptions);
+                if (checkerMethodName.startsWith(InstrumentationService.CHECK_METHOD_PREFIX)) {
+                    var checkerMethodArgumentTypes = Type.getArgumentTypes(checkerMethodDescriptor);
+                    var methodToInstrument = parseCheckerMethodSignature(checkerMethodName, checkerMethodArgumentTypes);
 
-                var checkerMethodArgumentTypes = Type.getArgumentTypes(checkerMethodDescriptor);
-                var methodToInstrument = parseCheckerMethodSignature(checkerMethodName, checkerMethodArgumentTypes);
-
-                var checkerParameterDescriptors = Arrays.stream(checkerMethodArgumentTypes).map(Type::getDescriptor).toList();
-                var checkMethod = new CheckMethod(Type.getInternalName(checkerClass), checkerMethodName, checkerParameterDescriptors);
-
-                methodsToInstrument.put(methodToInstrument, checkMethod);
+                    var checkerParameterDescriptors = Arrays.stream(checkerMethodArgumentTypes).map(Type::getDescriptor).toList();
+                    var checkMethod = new CheckMethod(Type.getInternalName(checkerClass), checkerMethodName, checkerParameterDescriptors);
 
+                    methodsToInstrument.put(methodToInstrument, checkMethod);
+                }
                 return mv;
             }
         };
         reader.accept(visitor, 0);
         return methodsToInstrument;
     }
 
+    @Override
+    public InstrumentationInfo lookupImplementationMethod(Class<?> implementationClass, Method instrumentMethod, Method checkMethod) {
+        checkMethodIsValid(implementationClass, instrumentMethod);
+
+        var instrumentMethodArguments = Type.getArgumentTypes(instrumentMethod);
+        var checkMethodArguments = Type.getArgumentTypes(checkMethod);
+
+        checkArguments(checkMethodArguments, instrumentMethodArguments, Modifier.isStatic(instrumentMethod.getModifiers()));
+
+        return new InstrumentationInfo(
+            new MethodKey(
+                Type.getInternalName(implementationClass),
+                instrumentMethod.getName(),
+                Arrays.stream(instrumentMethodArguments).map(Type::getInternalName).toList()
+            ),
+            new CheckMethod(
+                Type.getInternalName(checkMethod.getDeclaringClass()),
+                checkMethod.getName(),
+                Arrays.stream(checkMethodArguments).map(Type::getDescriptor).toList()
+            )
+        );
+    }
+
+    private static void checkArguments(Type[] checkMethodArguments, Type[] instrumentMethodArguments, boolean isStatic) {
```

Comment:
Naming is hard. I wonder if `instrumentMethodArguments` should be `targetMethodArguments`?

Reply:
I like target more - I have avoided it deliberately because I have a vague recollection we did not want to use it, but I'm all for it - I think it's easier to understand.

Reply:
Oh I do not have that recollection.

This project needs a glossary.

Reply:
Renamed to `targetMethodArguments`